## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [2]:
import pandas as pd

df = pd.read_csv('BootCampActivities/Challenges/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [3]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [4]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [5]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.count()

first          54639
last           54639
job            54639
phone          54639
email          54639
client_id      54639
order_id       54639
order_date     54639
order_week     54639
order_year     54639
item_id        54639
category       54639
subcategory    54639
unit_price     54639
unit_cost      54639
unit_weight    54639
qty            54639
line_number    54639
dtype: int64

In [7]:
# What three item categories had the most entries?
df["category"].value_counts()[0:3]

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [8]:
# For the category with the most entries, which subcategory had the most entries?
df_consumables=df.loc[df["category"]=='consumables']
df_consumables["subcategory"].value_counts()[0:1]

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [12]:
# Which five clients had the most entries in the data?
df["full_name"]=df["first"]+" "+df["last"]
df["full_name"].value_counts()[0:5]

full_name
Jessica Reyes      220
Angela Everett     211
Bryan Myers        209
Kendra Garrett     207
Alexandra Young    207
Name: count, dtype: int64

In [13]:
# Store the client ids of those top 5 clients in a list.
df=df.sort_values(by='client_id', key=lambda x: df["client_id"].value_counts()[x], ascending=False)
client_list=df["client_id"].unique()[0:5]
print(client_list)


[33615 66037 46820 38378 24741]


In [14]:
# How many total units (the qty column) did the client with the most entries order order?
df.loc[df["full_name"]=='Jessica Reyes',"qty"].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [15]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df["line_subtotal"]=df["unit_price"]*df["qty"]
df.head()

first   last                                           job  \
46459  Jessica  Reyes  Production designer, theatre/television/film   
6915   Jessica  Reyes  Production designer, theatre/television/film   
15706  Jessica  Reyes  Production designer, theatre/television/film   
16702  Jessica  Reyes  Production designer, theatre/television/film   
39133  Jessica  Reyes  Production designer, theatre/television/film   

            phone                     email  client_id  order_id  order_date  \
46459  5016915923  jessica.reyes@grimes.net      33615   8489769  2023-02-10   
6915   5016915923  jessica.reyes@grimes.net      33615   3122126  2023-04-29   
15706  5016915923  jessica.reyes@grimes.net      33615   7799056  2023-03-05   
16702  5016915923  jessica.reyes@grimes.net      33615   7221501  2023-04-10   
39133  5016915923  jessica.reyes@grimes.net      33615   8030004  2023-03-04   

       order_week  order_year         item_id     category        subcategory  \
46459           6        2023  XBM94536-93-1U        decor               misc   
6915           17        2023  BZD25887-03-4D  consumables  bathroom supplies   
15706           9        2023  ZBB37981-57-0A     software         accounting   
16702          15        2023  BAA80466-20-9X  consumables  bathroom supplies   
39133           9        2023  BBZ03090-18-0E     software         accounting   

       unit_price  unit_cost  unit_weight   qty  line_number      full_name  \
46459        0.87       0.50         0.96    22            3  Jessica Reyes   
6915        94.65      87.92         8.18    22            2  Jessica Reyes   
15706      122.15      76.81         1.17    97            5  Jessica Reyes   
16702       44.52      29.36         1.30  1120            1  Jessica Reyes   
39133      537.51     454.24         5.86   229            1  Jessica Reyes   

       line_subtotal  
46459          19.14  
6915         2082.30  
15706       11848.55  
16702       49862.40  
39133      123089.79

In [28]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def shipping(row):
    total_weight=row['unit_weight']*row['qty']
    if total_weight> 50:
        return total_weight*7
    return total_weight*10

df['shipping_price'] = df.apply(shipping, axis = 1)
df.head()

first   last                                           job  \
46459  Jessica  Reyes  Production designer, theatre/television/film   
6915   Jessica  Reyes  Production designer, theatre/television/film   
15706  Jessica  Reyes  Production designer, theatre/television/film   
16702  Jessica  Reyes  Production designer, theatre/television/film   
39133  Jessica  Reyes  Production designer, theatre/television/film   

            phone                     email  client_id  order_id  order_date  \
46459  5016915923  jessica.reyes@grimes.net      33615   8489769  2023-02-10   
6915   5016915923  jessica.reyes@grimes.net      33615   3122126  2023-04-29   
15706  5016915923  jessica.reyes@grimes.net      33615   7799056  2023-03-05   
16702  5016915923  jessica.reyes@grimes.net      33615   7221501  2023-04-10   
39133  5016915923  jessica.reyes@grimes.net      33615   8030004  2023-03-04   

       order_week  order_year  ... unit_cost unit_weight   qty  line_number  \
46459           6        2023  ...      0.50        0.96    22            3   
6915           17        2023  ...     87.92        8.18    22            2   
15706           9        2023  ...     76.81        1.17    97            5   
16702          15        2023  ...     29.36        1.30  1120            1   
39133           9        2023  ...    454.24        5.86   229            1   

           full_name  line_subtotal  shipping_price     line_price  line_cost  \
46459  Jessica Reyes          19.14          211.20     251.646450     222.20   
6915   Jessica Reyes        2082.30         1259.72    3651.156850    3193.96   
15706  Jessica Reyes       11848.55          794.43   13812.455650    8245.00   
16702  Jessica Reyes       49862.40        10192.00   65609.432000   43075.20   
39133  Jessica Reyes      123089.79         9393.58  144738.081725  113414.54   

        line_profit  
46459     29.446450  
6915     457.196850  
15706   5567.455650  
16702  22534.232000  
39133  31323.541725  

[5 rows x 24 columns]

In [17]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
def total_price(row):
    return (row['line_subtotal'] + row['shipping_price']) * 1.0925

df['line_price'] = df.apply(total_price, axis = 1)
df.head()

first   last                                           job  \
46459  Jessica  Reyes  Production designer, theatre/television/film   
6915   Jessica  Reyes  Production designer, theatre/television/film   
15706  Jessica  Reyes  Production designer, theatre/television/film   
16702  Jessica  Reyes  Production designer, theatre/television/film   
39133  Jessica  Reyes  Production designer, theatre/television/film   

            phone                     email  client_id  order_id  order_date  \
46459  5016915923  jessica.reyes@grimes.net      33615   8489769  2023-02-10   
6915   5016915923  jessica.reyes@grimes.net      33615   3122126  2023-04-29   
15706  5016915923  jessica.reyes@grimes.net      33615   7799056  2023-03-05   
16702  5016915923  jessica.reyes@grimes.net      33615   7221501  2023-04-10   
39133  5016915923  jessica.reyes@grimes.net      33615   8030004  2023-03-04   

       order_week  order_year  ...        subcategory unit_price unit_cost  \
46459           6        2023  ...               misc       0.87      0.50   
6915           17        2023  ...  bathroom supplies      94.65     87.92   
15706           9        2023  ...         accounting     122.15     76.81   
16702          15        2023  ...  bathroom supplies      44.52     29.36   
39133           9        2023  ...         accounting     537.51    454.24   

       unit_weight   qty  line_number      full_name  line_subtotal  \
46459         0.96    22            3  Jessica Reyes          19.14   
6915          8.18    22            2  Jessica Reyes        2082.30   
15706         1.17    97            5  Jessica Reyes       11848.55   
16702         1.30  1120            1  Jessica Reyes       49862.40   
39133         5.86   229            1  Jessica Reyes      123089.79   

      shipping_price     line_price  
46459         211.20     251.646450  
6915         1259.72    3651.156850  
15706         794.43   13812.455650  
16702       10192.00   65609.432000  
39133        9393.58  144738.081725  

[5 rows x 22 columns]

In [18]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost']=df["unit_cost"]*df["qty"]+df["shipping_price"]
df.head()


first   last                                           job  \
46459  Jessica  Reyes  Production designer, theatre/television/film   
6915   Jessica  Reyes  Production designer, theatre/television/film   
15706  Jessica  Reyes  Production designer, theatre/television/film   
16702  Jessica  Reyes  Production designer, theatre/television/film   
39133  Jessica  Reyes  Production designer, theatre/television/film   

            phone                     email  client_id  order_id  order_date  \
46459  5016915923  jessica.reyes@grimes.net      33615   8489769  2023-02-10   
6915   5016915923  jessica.reyes@grimes.net      33615   3122126  2023-04-29   
15706  5016915923  jessica.reyes@grimes.net      33615   7799056  2023-03-05   
16702  5016915923  jessica.reyes@grimes.net      33615   7221501  2023-04-10   
39133  5016915923  jessica.reyes@grimes.net      33615   8030004  2023-03-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
46459           6        2023  ...       0.87      0.50        0.96    22   
6915           17        2023  ...      94.65     87.92        8.18    22   
15706           9        2023  ...     122.15     76.81        1.17    97   
16702          15        2023  ...      44.52     29.36        1.30  1120   
39133           9        2023  ...     537.51    454.24        5.86   229   

       line_number      full_name  line_subtotal  shipping_price  \
46459            3  Jessica Reyes          19.14          211.20   
6915             2  Jessica Reyes        2082.30         1259.72   
15706            5  Jessica Reyes       11848.55          794.43   
16702            1  Jessica Reyes       49862.40        10192.00   
39133            1  Jessica Reyes      123089.79         9393.58   

          line_price  line_cost  
46459     251.646450     222.20  
6915     3651.156850    3193.96  
15706   13812.455650    8245.00  
16702   65609.432000   43075.20  
39133  144738.081725  113414.54  

[5 rows x 23 columns]

In [19]:
# Create a column for the profit of each line using line cost and line price
df['line_profit']=df["line_price"]-df["line_cost"]
df.head()

first   last                                           job  \
46459  Jessica  Reyes  Production designer, theatre/television/film   
6915   Jessica  Reyes  Production designer, theatre/television/film   
15706  Jessica  Reyes  Production designer, theatre/television/film   
16702  Jessica  Reyes  Production designer, theatre/television/film   
39133  Jessica  Reyes  Production designer, theatre/television/film   

            phone                     email  client_id  order_id  order_date  \
46459  5016915923  jessica.reyes@grimes.net      33615   8489769  2023-02-10   
6915   5016915923  jessica.reyes@grimes.net      33615   3122126  2023-04-29   
15706  5016915923  jessica.reyes@grimes.net      33615   7799056  2023-03-05   
16702  5016915923  jessica.reyes@grimes.net      33615   7221501  2023-04-10   
39133  5016915923  jessica.reyes@grimes.net      33615   8030004  2023-03-04   

       order_week  order_year  ... unit_cost unit_weight   qty  line_number  \
46459           6        2023  ...      0.50        0.96    22            3   
6915           17        2023  ...     87.92        8.18    22            2   
15706           9        2023  ...     76.81        1.17    97            5   
16702          15        2023  ...     29.36        1.30  1120            1   
39133           9        2023  ...    454.24        5.86   229            1   

           full_name  line_subtotal  shipping_price     line_price  line_cost  \
46459  Jessica Reyes          19.14          211.20     251.646450     222.20   
6915   Jessica Reyes        2082.30         1259.72    3651.156850    3193.96   
15706  Jessica Reyes       11848.55          794.43   13812.455650    8245.00   
16702  Jessica Reyes       49862.40        10192.00   65609.432000   43075.20   
39133  Jessica Reyes      123089.79         9393.58  144738.081725  113414.54   

        line_profit  
46459     29.446450  
6915     457.196850  
15706   5567.455650  
16702  22534.232000  
39133  31323.541725  

[5 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [22]:
# Check your work using the totals above

print(f"Order ID 2742071 had a total price of ${df.loc[df['order_id']==2742071,'line_price'].sum():,.2f}")
print(f"Order ID 2173913 had a total price of ${df.loc[df['order_id']==2173913,'line_price'].sum():,.2f}")
print(f"Order ID 6128929 had a total price of ${df.loc[df['order_id']==6128929,'line_price'].sum():,.2f}")
#some rounding on the last digit

Order ID 2742071 had a total price of $152,811.90
Order ID 2173913 had a total price of $162,388.72
Order ID 6128929 had a total price of $923,441.24


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [42]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
client_dict={}
i=0
for client in client_list:
    client_dict[i]={client : "$"+str(round(df.loc[df['client_id']==client,'line_price'].sum(),2))}
    i+=1
client_dict


{0: {33615: '$8377308.57'},
 1: {66037: '$10259514.8'},
 2: {46820: '$9743794.32'},
 3: {38378: '$12906550.88'},
 4: {24741: '$82268891.98'}}

In [43]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
summary_list=[]
for client in client_list:
    summary_list.append({"client_id" : client, "qty" : df.loc[df['client_id']==client,'qty'].sum(),
                    "shipping_price" : df.loc[df['client_id']==client,'shipping_price'].sum(),
                    "line_price" : df.loc[df['client_id']==client,'line_price'].sum(),
                    "line_cost" : df.loc[df['client_id']==client,'line_cost'].sum(),
                    "line_profit" : df.loc[df['client_id']==client,'line_profit'].sum()})
summary_df=pd.DataFrame(summary_list)
summary_df

client_id     qty  shipping_price    line_price    line_cost   line_profit
0      33615   64313      1828984.89  8.377309e+06   6175313.91  2.201995e+06
1      66037   43018      1395151.85  1.025951e+07   7004482.98  3.255032e+06
2      46820   75768      1601448.84  9.743794e+06   7007191.64  2.736603e+06
3      38378   73667      3429455.40  1.290655e+07   9634720.98  3.271830e+06
4      24741  239862      5126448.37  8.226889e+07  45688899.71  3.657999e+07

In [44]:
# Format the data and rename the columns to names suitable for presentation.
renamed_df=summary_df.rename(columns={"client_id" : "Client ID","qty" : "Units","shipping_price" : "Shipping(millions)",
                                      "line_price" : "Total Revenue(millions)","line_cost" : "Total Cost(millions)",
                                      "line_profit" : "Total Profit(millions)"})
# Define the money columns. 
money_columns=["Shipping(millions)","Total Revenue(millions)","Total Cost(millions)","Total Profit(millions)"]

# Define a function that converts a dollar amount to millions.
def currency_format_millions(money):
    return money/1000000

# Apply the currency_format_millions function to only the money columns. 
renamed_df[money_columns] = renamed_df[money_columns].apply(currency_format_millions)

# Rename the columns to reflect the change in the money format. 
renamed_df



Client ID   Units  Shipping(millions)  Total Revenue(millions)  \
0      33615   64313            1.828985                 8.377309   
1      66037   43018            1.395152                10.259515   
2      46820   75768            1.601449                 9.743794   
3      38378   73667            3.429455                12.906551   
4      24741  239862            5.126448                82.268892   

   Total Cost(millions)  Total Profit(millions)  
0              6.175314                2.201995  
1              7.004483                3.255032  
2              7.007192                2.736603  
3              9.634721                3.271830  
4             45.688900               36.579992

In [45]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
renamed_df.sort_values(by="Total Profit(millions)", ascending=False)

Client ID   Units  Shipping(millions)  Total Revenue(millions)  \
4      24741  239862            5.126448                82.268892   
3      38378   73667            3.429455                12.906551   
1      66037   43018            1.395152                10.259515   
2      46820   75768            1.601449                 9.743794   
0      33615   64313            1.828985                 8.377309   

   Total Cost(millions)  Total Profit(millions)  
4             45.688900               36.579992  
3              9.634721                3.271830  
1              7.004483                3.255032  
2              7.007192                2.736603  
0              6.175314                2.201995

In [ ]:
#The client with the largest amount of orders was not the client with the largest spending.
#The client with the least amount of orders (from the Top 5) had the largest spending and units.
#The client with the second most amount of orders had the smallest amount in shipping cost.